tools/cxbe/Xbe.cpp

```diff
@@ -1527,3 +1524,91 @@ uint08 *Xbe::GetLogoBitmap(uint32 x_dwSize)
+    *extra_bytes = 0;
+    uint32 idtRVA = x_Exe->m_OptionalHeader.m_image_data_directory[IMAGE_DIRECTORY_ENTRY_IMPORT].m_virtual_addr;
+    if(idtRVA == 0) {
+        return 0;
+    }
+
+    uint32 count = 0;
+    const IMAGE_IMPORT_DESCRIPTOR *descr = reinterpret_cast<const IMAGE_IMPORT_DESCRIPTOR *>(x_Exe->ReadAddr(idtRVA));
+    for (; descr->FirstThunk; ++descr) {
+        uint32 nameRVA = descr->Name;
+        const char *name = reinterpret_cast<const char *>(x_Exe->ReadAddr(nameRVA));
+        if (strcasecmp(name, kKernelImageName) != 0) {
+            *extra_bytes += (strlen(name) + 1) * 2;
+            ++count;
+        }
+    }
+
+    return count;
+}
+
+// sets the dwKernelImageThunkAddr and dwNonKernelImportDirAddr fields from the given Exe.
+bool Xbe::ProcessImportTable(class Exe *x_Exe, bool x_bRetail) {
+    uint32 idtRVA = x_Exe->m_OptionalHeader.m_image_data_directory[IMAGE_DIRECTORY_ENTRY_IMPORT].m_virtual_addr;
+    if(idtRVA == 0) {
+        printf("Xbe::ProcessImportTable: No import directory found, exe may be invalid.\n");
+        return true;
+    }
+
+    uint32 kernel_thunk_addr = 0;
+    const IMAGE_IMPORT_DESCRIPTOR *peDescriptor = reinterpret_cast<const IMAGE_IMPORT_DESCRIPTOR *>(GetAddr(idtRVA + m_Header.dwPeBaseAddr));
+    XBE_IMAGE_IMPORT_DESCRIPTOR *xbeDescriptor = nullptr;
+    char *wideCharName = nullptr;
+    uint32 nextImportNameAddr = m_ImportNameAddr;
+    if (m_Header.dwNonKernelImportDirAddr) {
+        xbeDescriptor = reinterpret_cast<XBE_IMAGE_IMPORT_DESCRIPTOR *>(GetAddr(m_Header.dwNonKernelImportDirAddr));
+        if (!xbeDescriptor) {
+            SetError("Xbe::ProcessImportTable: Non kernel imports exist but target descriptor table is null", true);
+            return false;
+        }
+
+        wideCharName = reinterpret_cast<char*>(GetAddr(m_ImportNameAddr));
+        if (!wideCharName) {
+            SetError("Xbe::ProcessImportTable: Non kernel imports exist but target name table is null", true);
+            return false;
+        }
+    }
+
+    for (int32 i = 0; peDescriptor->FirstThunk; ++i, ++peDescriptor) {
+        uint32 nameRVA = peDescriptor->Name;
+        const char *name = reinterpret_cast<const char *>(GetAddr(nameRVA + m_Header.dwPeBaseAddr));
+        if (!strcasecmp(name, kKernelImageName)) {
+            kernel_thunk_addr = peDescriptor->FirstThunk + m_Header.dwPeBaseAddr;
+        } else {
+            xbeDescriptor->FirstThunk = peDescriptor->FirstThunk + m_Header.dwPeBaseAddr;
+            xbeDescriptor->WideCharName = nextImportNameAddr;
+
+            size_t len = strlen(name) + 1;
+            for (uint32 ch = 0; ch < len; ++ch) {
+                *wideCharName++ = name[ch];
+                *wideCharName++ = 0;
+                nextImportNameAddr += 2;
+            }
+            ++xbeDescriptor;
+        }
+    }
+
+    if (kernel_thunk_addr == 0) {
+        printf("Xbe::ProcessImportTable: No imports from xboxkrnl.exe, exe may be invalid.\n");
```

Comment:
This should use `%s` and `kKernelImageName` so this doesn't break when modifying `kKernelImageName`.

Reply:
Done.